### PR TITLE
Vb/pdgone

### PR DESCRIPTION
**Pull request by vineetbansal**
Stripped off pandas from the rest of the codebase.

### CODE REVIEW

**Comment by vineetbansal** [comment]
A lot of the acrobatics happening in this branch have to do with the fact that:
- some code is accessing the `_metadata` attribute directly and is thus tied to its implementation.
- It's a bit tricky to get the nth row of each value of a dict of lists, merge two such dicts etc, hence some code with dict comprehension, `zip` etc.
- The relion interop code was written to differentiate between scalar values and vectors on the basis of the incoming object being a `DataFrame`. Now that they're all dicts, one has to peek at the values.

Probably a good project for someone down the line would be to have a `Metadata` class to contain some of the ugliness here. On the other hand, the relion code is dealing with general-purpose 2D data and is not relying on it coming from metadata of an `ImageSource`, so gains might be limited.

I've submitted this as a separate PR so it can be merged after `vb/src_ascopy` is merged, or before - depending on how much disruption is acceptable, and the resources available to review and test these more thoroughly.

File: src/aspire/storage/starfile.py
Comment (garrettwrong) [inline src/aspire/storage/starfile.py]: Debating which is easier to read, `list(map(str, _v))`

File: src/aspire/storage/starfile.py
Comment (garrettwrong) [inline src/aspire/storage/starfile.py:145]: If one of the values is not iterable, I think we might not hit the assert check below (L155).

Is this a case of the prior np array keeping the size enforced behind the scenes? Assuming so this is probably safe enough.

File: src/aspire/source/coordinates.py
Comment (garrettwrong) [inline src/aspire/source/coordinates.py:230]: 226 and 227 can be combined is `_box_coord_from_center` is generalized to multiple centers.  I'll make a followup issue.

**Review by garrettwrong — APPROVED** [review]
Thanks @vineetbansal , tried running this against a known bad file (10028 shiny particles), and ASPIRE still seems to indicate it is bad, so I'm happy with that.  I think this is probably ready for Joakim to also sign off on.

There were a few small things, but I think they are just questions and minor cosmetic followups we can do. Thanks!

File: src/aspire/storage/starfile.py
Comment (vineetbansal) [reply src/aspire/storage/starfile.py]: Yea - too many underscores on this one. I'll change that.

File: src/aspire/storage/starfile.py
Comment (vineetbansal) [reply src/aspire/storage/starfile.py:145]: I was thinking if the incoming block has some iterables and some scalars, then everything is hosed anyway when trying to create a `loop_` of out it, so no point in ensuring that whatever iterables remain are the same length. Probably one good check here might be to make sure that either all of them are iterables, or none of them are, and set the `multiple_rows` variable after that check.

The incoming dict has no enforcement on the length of its individual values. If it's coming from metadata (from `get_metadata` with `as_dict` either True or False), then its rectangular anyway so its safe.

Comment (garrettwrong) [reply src/aspire/storage/starfile.py:145]: Seems reasonable, thanks!

**Comment by garrettwrong** [comment]
Looks like the redundant PR merge conflicted themselves.

File: src/aspire/storage/starfile.py
Comment (vineetbansal) [reply src/aspire/storage/starfile.py]: changed

**Comment by garrettwrong** [comment]
Docs build failed because EMDB ftp was offline yesterday ( and so we could not download our datafile). Its online now, restarting jobs.

File: setup.py
Comment (janden) [inline setup.py:36]: :stuck_out_tongue_closed_eyes:

File: src/aspire/source/coordinates.py
Comment (janden) [inline src/aspire/source/coordinates.py]: Cleaner way to do this? `list(data_block.values())[0]`? Maybe just as ugly…

File: src/aspire/source/coordinates.py
Comment (janden) [inline src/aspire/source/coordinates.py]: Think `np.stack` suffices here. Also maybe split call up into two lines (one to construct the raw params, one to get the unique indices).

File: src/aspire/source/coordinates.py
Comment (janden) [inline src/aspire/source/coordinates.py]: `df` → `data_block`

File: src/aspire/source/relion.py
Comment (janden) [inline src/aspire/source/relion.py:111]: Again, maybe split this up into two lines.

File: src/aspire/source/relion.py
Comment (janden) [inline src/aspire/source/relion.py:222]: Since `indices` is now supplied as an argument, we probably don't need to return it anymore.

File: src/aspire/storage/starfile.py
Comment (janden) [inline src/aspire/storage/starfile.py]: Maybe `try: iter(_v)` is more appropriate here?

File: src/aspire/storage/starfile.py
Comment (janden) [inline src/aspire/storage/starfile.py:21]: Comment briefly on what this is supposed to do? Some sort of normalization of block data?

File: src/aspire/storage/starfile.py
Comment (janden) [inline src/aspire/storage/starfile.py]: `[str(x[i]) for x in loop_data]`?

**Review by janden — CHANGES_REQUESTED** [review]
Much cleaner, although I confess there are some list comprehension acrobatics here that I have trouble following (in the  `aspire.source.*` modules). Perhaps some can be simplified (or commented) but this is not too urgent to hold up a merge.

File: src/aspire/source/coordinates.py
Comment (garrettwrong) [reply src/aspire/source/coordinates.py]: Yeah, surprised I missed that one. Changed.

File: src/aspire/source/coordinates.py
Comment (garrettwrong) [reply src/aspire/source/coordinates.py]: Refactored

File: src/aspire/source/relion.py
Comment (garrettwrong) [reply src/aspire/source/relion.py:111]: refactored.

File: src/aspire/source/coordinates.py
Comment (garrettwrong) [reply src/aspire/source/coordinates.py]: replaced

File: src/aspire/source/relion.py
Comment (garrettwrong) [reply src/aspire/source/relion.py:222]: I think it is actually unpacked in the future, and would be needed then , unless we keep track of the indices in another structure.
Good idea, but i'm not willing to touch that threadpool code in a rush. (it has no reason to exist, but that is another matter entirely).  Will create an issue to delete it.

Comment (garrettwrong) [reply src/aspire/source/relion.py:222]: https://github.com/ComputationalCryoEM/ASPIRE-Python/issues/922

File: src/aspire/storage/starfile.py
Comment (garrettwrong) [reply src/aspire/storage/starfile.py]: yeah, sorry I missed that one too, replaced

File: src/aspire/storage/starfile.py
Comment (garrettwrong) [reply src/aspire/storage/starfile.py:21]: # The following code repacks provided blocks as strings while handling both singleton and iterable inputs.

File: src/aspire/storage/starfile.py
Comment (garrettwrong) [reply src/aspire/storage/starfile.py]: I will take a look at this a bit more, but I think will result in more code.

Comment (garrettwrong) [reply src/aspire/storage/starfile.py]: Only slightly more code, but I added comments so perhaps it is clearer now.

File: src/aspire/source/coordinates.py
Comment (janden) [reply src/aspire/source/coordinates.py]: Thanks.

File: src/aspire/source/relion.py
Comment (janden) [reply src/aspire/source/relion.py:222]: Yeah I saw that, so issue is probably the way to go. Thanks.

File: src/aspire/storage/starfile.py
Comment (janden) [reply src/aspire/storage/starfile.py]: This seems to happen often enough in this new code that having a utility `is_iterable` may be appropriate here?

File: src/aspire/storage/starfile.py
Comment (janden) [reply src/aspire/storage/starfile.py]: Thanks!

File: src/aspire/storage/starfile.py
Comment (janden) [reply src/aspire/storage/starfile.py]: Can make an issue.

**Review by janden — APPROVED** [review]
Looks good.

File: src/aspire/storage/starfile.py
Comment (garrettwrong) [reply src/aspire/storage/starfile.py]: Sure https://github.com/ComputationalCryoEM/ASPIRE-Python/issues/925

**Pull request merged**
